### PR TITLE
Python: Don't install deps by default for all users

### PR DESCRIPTION
**Pull request by RasmusWL**
### Merge / deployment checklist

- [ ] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [x] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

**Comment by RasmusWL** [comment]
I think we can ignore `Swift analysis using a custom build command` as a spurious failure :blush:

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ See the [releases page](https://github.com/github/codeql-action/releases) for th
 
 ## [UNRELEASED]
 
-No user facing changes.
+- We have disabled Python dependency installation for all users by default. This improves the speed of analysis while having only a very minor impact on results. You can override this behavior until CodeQL CLI 2.17.0 is released by setting `CODEQL_ACTION_DISABLE_PYTHON_DEPENDENCY_INSTALLATION=false` in your workflow. [#2031](https://github.com/github/codeql-action/pull/2031)
```
Comment (aeisenberg) [inline CHANGELOG.md:7]: I don't think it's safe to give a precise version when we will remove this feature. How about this:

```suggestion
- We have disabled Python dependency installation for all users by default. This improves the speed of analysis while having only a very minor impact on results. You can override this behavior by setting `CODEQL_ACTION_DISABLE_PYTHON_DEPENDENCY_INSTALLATION=false` in your workflow. This feature will be removed in future versions of the CodeQL Action. [#2031](https://github.com/github/codeql-action/pull/2031)
```

When this feature is finally removed in 2.17.0, we can add another changelog entry.

Comment (RasmusWL) [reply CHANGELOG.md:7]: fine by me, thanks :heart:

**Comment by aeisenberg** [comment]
LGTM! I'll avoid approving this until you move this out of draft mode.

**Review by iamannalynleynes1206 — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,7 @@ Note that the only difference between `v2` and `v3` of the CodeQL Action is the
 
 ## [UNRELEASED]
 
+- We are rolling out a feature in January 2024 that will disable Python dependency installation by default for all users. This improves the speed of analysis while having only a very minor impact on results. You can override this behavior by setting `CODEQL_ACTION_DISABLE_PYTHON_DEPENDENCY_INSTALLATION=false` in your workflow, however we plan to remove this ability in future versions of the CodeQL Action. [#2031](https://github.com/github/codeql-action/pull/2031)
```
Comment (henrymercer) [inline CHANGELOG.md:9]: [Non-blocking] Currently we disable Python dependency installation for CodeQL v2.16.0 and later, however the Action supports CodeQL versions all the way back to 2.11.6.  Do we plan to keep supporting Python dependency installation for old CLIs in the Action until support for v2.15.5 is deprecated in about a year from now?

File: src/init-action.ts [hunk 3]
```diff
@@ -447,15 +451,24 @@ async function run() {
...
     ) {
       core.exportVariable(
         "CODEQL_EXTRACTOR_PYTHON_DISABLE_LIBRARY_EXTRACTION",
         "true",
       );
+    } else {
+      core.exportVariable(
+        "CODEQL_EXTRACTOR_PYTHON_FORCE_ENABLE_LIBRARY_EXTRACTION_UNTIL_2_17_0",
+        "true",
+      );
```
Comment (henrymercer) [inline src/init-action.ts:471]: I think I can have a good guess at what this does (I'm guessing library extraction is disabled by default in CodeQL v2.16.0) but could you add a comment just to make it crystal clear?  Thanks!

**Review by henrymercer — COMMENTED** [review]
This looks great, just a couple of comments.

File: CHANGELOG.md (hunk 2 above)
Comment (RasmusWL) [reply CHANGELOG.md:9]: I had hoped to be able delete the `python-setup` folder and all the logic for dependency installation soon, but let's discuss this aspect some more :+1:

File: src/init-action.ts (hunk 3 above)
Comment (RasmusWL) [reply src/init-action.ts:459]: Fixed in afef25e1e74d7376709cd9521c5c6dcf55dceba5

File: src/analyze.ts [hunk 4]
```diff
@@ -105,10 +105,14 @@ async function setupPythonExtractor(
...
-    await features.getValue(
+    (await features.getValue(
       Feature.DisablePythonDependencyInstallationEnabled,
       codeql,
-    )
+    )) ||
+    (await features.getValue(
+      Feature.PythonDefaultIsToSkipDependencyInstallationEnabled,
+      codeql,
+    ))
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply src/analyze.ts]: Fixed in 2e27b3c56bdb1864f5dd58da25fee61f6cd7bb81

File: CHANGELOG.md (hunk 2 above)
Comment (henrymercer) [reply CHANGELOG.md:9]: 👍 We can always start applying this to earlier CLI versions later on.  Happy to discuss!

**Review by henrymercer — APPROVED** [review]
Great, thank you!  FYI, if after pushing commits you realize that you forgot to rebuild the Action, you can now apply the "Rebuild" label to have Actions rebuild it for you.

**Pull request merged**